go/teleportr/api/server.go

```diff
@@ -0,0 +1,400 @@
+
+	defaultTimeout = 10 * time.Second
+)
+
+type Server struct {
+	ctx              context.Context
+	l1Client         *ethclient.Client
+	depositAddr      common.Address
+	numConfirmations uint64
+	depositContract  *deposit.TeleportrDeposit
+	database         *db.Database
+}
+
+type StatusResponse struct {
+	CurrentBalanceWei   string `json:"current_balance_wei"`
+	MaximumBalanceWei   string `json:"maximum_balance_wei"`
+	MinDepositAmountWei string `json:"min_deposit_amount_wei"`
+	MaxDepositAmountWei string `json:"max_deposit_amount_wei"`
+	IsAvailable         bool   `json:"is_available"`
+}
+
+func (s *Server) HandleStatus(w http.ResponseWriter, req *http.Request) {
+	ctx, cancel := context.WithTimeout(s.ctx, defaultTimeout)
+	defer cancel()
+
+	maxBalance, err := s.depositContract.MaxBalance(&bind.CallOpts{
+		Context: ctx,
+	})
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	minDepositAmount, err := s.depositContract.MinDepositAmount(&bind.CallOpts{
+		Context: ctx,
+	})
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	maxDepositAmount, err := s.depositContract.MaxDepositAmount(&bind.CallOpts{
+		Context: ctx,
+	})
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	curBalance, err := s.l1Client.BalanceAt(ctx, s.depositAddr, nil)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	balanceAfterMaxDeposit := new(big.Int).Add(
+		curBalance, maxDepositAmount,
+	)
+	isAvailable := curBalance.Cmp(balanceAfterMaxDeposit) >= 0
+
+	resp := StatusResponse{
+		CurrentBalanceWei:   curBalance.String(),
+		MaximumBalanceWei:   maxBalance.String(),
+		MinDepositAmountWei: minDepositAmount.String(),
+		MaxDepositAmountWei: maxDepositAmount.String(),
+		IsAvailable:         isAvailable,
+	}
+
+	jsonResp, err := json.Marshal(resp)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set(ContentTypeHeader, ContentTypeJSON)
+	_, _ = w.Write(jsonResp)
+}
+
+type EstimateResponse struct {
+	BaseFee     string `json:"base_fee"`
+	GasTipCap   string `json:"gas_tip_cap"`
+	GasFeeCap   string `json:"gas_fee_cap"`
+	GasEstimate string `json:"gas_estimate"`
+}
+
+func (s *Server) HandleEstimate(w http.ResponseWriter, req *http.Request) {
+	amounts, ok := req.URL.Query()["amount"]
+	if !ok || len(amounts) < 1 {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	amount, ok := new(big.Int).SetString(amounts[0], 10)
+	if !ok {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	addresses, ok := req.URL.Query()["address"]
+	if !ok || len(addresses) < 1 {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	address := common.HexToAddress(addresses[0])
+
+	ctx, cancel := context.WithTimeout(s.ctx, defaultTimeout)
+	defer cancel()
+
+	gasTipCap, err := s.l1Client.SuggestGasTipCap(ctx)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
We should log all these errors - right now they'll be swallowed.

Reply:
I found that wrapping the handlers in an `ErrorTrackingHandler` that logs errors and writes `http.StatusInternalServerError` is a clean way to do it. This also provides a hook for you to capture error metrics as well.

Reply:
added! thanks for the suggestion 👍